src/strategy/actions/LootRollAction.cpp

```diff
@@ -5,151 +5,2088 @@
+            {
+                // Poor (grey) or Common (white) gear -> considered "bad"
+                hasVeryBadItem = true;
+            }
+
+            if (hasVeryBadItem)
+            {
+                LOG_DEBUG("playerbots",
+                          "[LootRollDBG] cross-armor: desperate upgrade candidate {} blocks cross-armor NEED on item={} \"{}\"",
+                          member->GetName(), proto->ItemId, proto->Name1);
+                return true;
+            }
+
+            return false;
+        });
+}
+
+// True if some bot has this jewelry/cloak as an upgrade and an empty/very bad item in the corresponding slots.
+static bool GroupHasDesperateJewelryUpgradeUser(Player* self, ItemTemplate const* proto, int32 randomProperty)
+{
+    if (!self || !proto)
+    {
+        return false;
+    }
+
+    uint8 jewelrySlots[2];
+    uint8 slotsCount = 0;
+
+    switch (proto->InventoryType)
+    {
+        case INVTYPE_NECK:
+            jewelrySlots[0] = EQUIPMENT_SLOT_NECK;
+            slotsCount = 1;
+            break;
+        case INVTYPE_FINGER:
+            jewelrySlots[0] = EQUIPMENT_SLOT_FINGER1;
+            jewelrySlots[1] = EQUIPMENT_SLOT_FINGER2;
+            slotsCount = 2;
+            break;
+        case INVTYPE_TRINKET:
+            jewelrySlots[0] = EQUIPMENT_SLOT_TRINKET1;
+            jewelrySlots[1] = EQUIPMENT_SLOT_TRINKET2;
+            slotsCount = 2;
+            break;
+        case INVTYPE_CLOAK:
+            jewelrySlots[0] = EQUIPMENT_SLOT_BACK;
+            slotsCount = 1;
+            break;
+        default:
+            // Not jewelry/cloak -> nothing to do here
+            return false;
+    }
+
+    std::string const param = BuildItemUsageParam(proto->ItemId, randomProperty);
+
+    return ForEachBotGroupMember(self,
+        [&](Player* member, PlayerbotAI* memberAI) -> bool
+        {
+            AiObjectContext* ctx = memberAI->GetAiObjectContext();
+            if (!ctx)
+                return false;
+
+            ItemUsage usage = ctx->GetValue<ItemUsage>("item usage", param)->Get();
+            if (usage != ITEM_USAGE_EQUIP && usage != ITEM_USAGE_REPLACE)
+                return false; // not a true upgrade for this bot
+
+            ItemTemplate const* bestProto = nullptr;
+
+            // Look at the best currently equipped jewelry in the corresponding slots
+            for (uint8 i = 0; i < slotsCount; ++i)
+            {
+                uint8 const slot = jewelrySlots[i];
+                Item* oldItem = member->GetItemByPos(INVENTORY_SLOT_BAG_0, slot);
+                if (!oldItem)
+                    continue;
+
+                ItemTemplate const* oldProto = oldItem->GetTemplate();
+                if (!oldProto)
+                    continue;
+
+                if (!bestProto || oldProto->ItemLevel > bestProto->ItemLevel)
+                    bestProto = oldProto;
+            }
+
+            bool hasVeryBadItem = false;
+
+            if (!bestProto)
```

Comment:
Why is there an if else here when they can be checked with a regular if? It sets the same variable based on it so why not have it like so
```cpp
if (!bestproto || bestProto->Quality <= ITEM_QUALITY_NORMAL)
```

The are also curly brackets with a single line if that shouldn't be there